cmd/thanos/sidecar.go

```diff
@@ -195,24 +215,53 @@ func runSidecar(
 		})
 	}
 
+	var (
+		bkt     objstore.Bucket
+		bucket  string
+		closeFn      = func() {}
+		uploads bool = true
+	)
+
+	s3Config := &s3.Config{
+		Bucket:    s3Bucket,
+		Endpoint:  s3Endpoint,
+		AccessKey: s3AccessKey,
+		SecretKey: s3SecretKey,
+		Insecure:  s3Insecure,
+	}
+
+	// The background shipper continuously scans the data directory and uploads
+	// new blocks to Google Cloud Storage or an S3-compatible storage service.
 	if gcsBucket != "" {
-		// The background shipper continuously scans the data directory and uploads
-		// new found blocks to Google Cloud Storage.
 		gcsClient, err := storage.NewClient(context.Background())
 		if err != nil {
 			return errors.Wrap(err, "create GCS client")
 		}
 
-		var bkt objstore.Bucket
 		bkt = gcs.NewBucket(gcsBucket, gcsClient.Bucket(gcsBucket), reg)
-		bkt = objstore.BucketWithMetrics(gcsBucket, bkt, reg)
+		closeFn = func() { gcsClient.Close() }
```

Comment:
you can assign like `closeFn = gsClient.Close` directly. Or make `closer io.Closer` and assign `gcsClient`.

Reply:
Good point @fabxc. Sorry @TimSimmons, I suggested `closeFn = func() { gcsClient.Close() }` (:

Reply:
Ah yes that's better. I think I'll keep it as a `func()` over an `io.Closer` in case some driver has a function they'd like to call that isn't `Close() err`. But I'll leave a comment to make it clear what the purpose of `closeFn` is.

Reply:
Eh nvm, if that situation comes up, someone can make it so. Using an `io.Closer` is nicer.

Reply:
Well. Actually that means I'd have to instantiate an `io.Closer` in the variable definition, which is not quite trivial. I'll leave it as is.